<a href="https://colab.research.google.com/github/vanshdhiman090/-Portuguese-bank-telemarketing-campaign-/blob/main/Case_Study_bank_campaign_cleaning.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

For solving this Business Problem  i will use Google data analytics mehtod which has six Phases.

> Ask

> Prepare

> Process

> Analyze

> Share

> Act

# Ask Phase

**Business Problem:**

*A Portuguese bank conducted a telemarketing campaign to sell term deposit subscriptions to existing customers. The campaign generated 45,211 customer interactions. Leadership needs to understand which customer profiles and campaign strategies drove conversions — and which wasted resources.*

**Goal:**

*Identify the key drivers of term deposit subscription using customer demographic and campaign interaction data. Deliver actionable recommendations to improve conversion rate and reduce cost-per-acquisition in future campaigns.*

**Stakeholder:**

*Marketing Director and Campaign Strategy Team.*

**Success Metric:**

*A dashboard that clearly identifies high-converting customer segments and campaign behaviours — with at least 3 actionable recommendations.*

**Business Questions:**


*   What was the overall conversion rate of the campaign — and is it acceptable?

*   Which customer segments (age, job, education) converted at the highest rate?


*   Does call duration and number of contacts impact conversion?

*   Did customers with a successful previous campaign convert again?


*   Which months and contact methods drove the best results?

# Prepare Phase

In [ ]:
# Importing libraries
import pandas as pd
import numpy as np


In [ ]:
df = pd.read_csv('/content/bank-additional-full.csv' , sep= ';')
df = pd.DataFrame(df)


In [ ]:
print(df.shape)

(41188, 21)


In [ ]:
# Columns Data type
print(df.dtypes)

age                 int64
job                object
marital            object
education          object
default            object
housing            object
loan               object
contact            object
month              object
day_of_week        object
duration            int64
campaign            int64
pdays               int64
previous            int64
poutcome           object
emp.var.rate      float64
cons.price.idx    float64
cons.conf.idx     float64
euribor3m         float64
nr.employed       float64
y                  object
dtype: object


**Data Sources & Location**

Source Identification: External (Kaggle)

Data Storage: Document Download section name by (Case study Power bi bank-additional-full)

Accessibility: Open-source , freely avalable on Kaggle.

**ROCCC Framework**

Who collected this data? - Portuguese banking institution (Kaggle)

Is it reliable?  - Yes — 41,188 rows and 21 cokumns

Original - Yes — created by Portuguese banking institution

Comprhensive - Mostly yes

Current - No  (collected in 2011)

Cited - Yes (UCI Bank Marketing Dataset on kaggel)




**Data  Credibility**

*Dataset passes ROCCC criteria. Key limitations — data age (2008–2010) and single-institution scope — are acknowledged and do not affect the analytical framework applied.*

**Data Ethics and Privacy**

*Data was anonymized prior to publication. No individual consent mechanism is documented. Used strictly for educational purposes.*

**Bias Assessment**
1. Survivorship Bias — only contacted customers included
2. Duration Bias — call duration known only after call ends
3. Time Period Bias — data from 2008–2010 financial crisis
4. Class Imbalance — check ratio of yes vs no in y column

# Process Phase

In [ ]:
# Data integrity process

df_raw = df.copy()
print(display(df))

,age,job,marital,education,default,housing,loan,contact,month,day_of_week,...,campaign,pdays,previous,poutcome,emp.var.rate,cons.price.idx,cons.conf.idx,euribor3m,nr.employed,y
0,56,housemaid,married,basic.4y,no,no,no,telephone,may,mon,...,1,999,0,nonexistent,1.1,93.994,-36.4,4.857,5191.0,no
1,57,services,married,high.school,unknown,no,no,telephone,may,mon,...,1,999,0,nonexistent,1.1,93.994,-36.4,4.857,5191.0,no
2,37,services,married,high.school,no,yes,no,telephone,may,mon,...,1,999,0,nonexistent,1.1,93.994,-36.4,4.857,5191.0,no
3,40,admin.,married,basic.6y,no,no,no,telephone,may,mon,...,1,999,0,nonexistent,1.1,93.994,-36.4,4.857,5191.0,no
4,56,services,married,high.school,no,no,yes,telephone,may,mon,...,1,999,0,nonexistent,1.1,93.994,-36.4,4.857,5191.0,no
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
41183,73,retired,married,professional.course,no,yes,no,cellular,nov,fri,...,1,999,0,nonexistent,-1.1,94.767,-50.8,1.028,4963.6,yes
41184,46,blue-collar,married,professional.course,no,no,no,cellular,nov,fri,...,1,999,0,nonexistent,-1.1,94.767,-50.8,1.028,4963.6,no
41185,56,retired,married,university.degree,no,yes,no,cellular,nov,fri,...,2,999,0,nonexistent,-1.1,94.767,-50.8,1.028,4963.6,no
41186,44,technician,married,professional.course,no,no,no,cellular,nov,fri,...,1,999,0,nonexistent,-1.1,94.767,-50.8,1.028,4963.6,yes


None


In [ ]:
print(df.shape)

(41188, 21)


In [ ]:
print(df.info())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 41188 entries, 0 to 41187
Data columns (total 21 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   age             41188 non-null  int64  
 1   job             41188 non-null  object 
 2   marital         41188 non-null  object 
 3   education       41188 non-null  object 
 4   default         41188 non-null  object 
 5   housing         41188 non-null  object 
 6   loan            41188 non-null  object 
 7   contact         41188 non-null  object 
 8   month           41188 non-null  object 
 9   day_of_week     41188 non-null  object 
 10  duration        41188 non-null  int64  
 11  campaign        41188 non-null  int64  
 12  pdays           41188 non-null  int64  
 13  previous        41188 non-null  int64  
 14  poutcome        41188 non-null  object 
 15  emp.var.rate    41188 non-null  float64
 16  cons.price.idx  41188 non-null  float64
 17  cons.conf.idx   41188 non-null 

In [ ]:
print(df.describe())

               age      duration      campaign         pdays      previous  \
count  41188.00000  41188.000000  41188.000000  41188.000000  41188.000000   
mean      40.02406    258.285010      2.567593    962.475454      0.172963   
std       10.42125    259.279249      2.770014    186.910907      0.494901   
min       17.00000      0.000000      1.000000      0.000000      0.000000   
25%       32.00000    102.000000      1.000000    999.000000      0.000000   
50%       38.00000    180.000000      2.000000    999.000000      0.000000   
75%       47.00000    319.000000      3.000000    999.000000      0.000000   
max       98.00000   4918.000000     56.000000    999.000000      7.000000   

       emp.var.rate  cons.price.idx  cons.conf.idx     euribor3m   nr.employed  
count  41188.000000    41188.000000   41188.000000  41188.000000  41188.000000  
mean       0.081886       93.575664     -40.502600      3.621291   5167.035911  
std        1.570960        0.578840       4.628198    

In [ ]:
print(df.isnull().sum())

age               0
job               0
marital           0
education         0
default           0
housing           0
loan              0
contact           0
month             0
day_of_week       0
duration          0
campaign          0
pdays             0
previous          0
poutcome          0
emp.var.rate      0
cons.price.idx    0
cons.conf.idx     0
euribor3m         0
nr.employed       0
y                 0
dtype: int64


In [ ]:
print(df.duplicated().sum())

12


In [ ]:
print(df.drop_duplicates())
df = df.drop_duplicates()

       age          job  marital            education  default housing loan  \
0       56    housemaid  married             basic.4y       no      no   no   
1       57     services  married          high.school  unknown      no   no   
2       37     services  married          high.school       no     yes   no   
3       40       admin.  married             basic.6y       no      no   no   
4       56     services  married          high.school       no      no  yes   
...    ...          ...      ...                  ...      ...     ...  ...   
41183   73      retired  married  professional.course       no     yes   no   
41184   46  blue-collar  married  professional.course       no      no   no   
41185   56      retired  married    university.degree       no     yes   no   
41186   44   technician  married  professional.course       no      no   no   
41187   74      retired  married  professional.course       no     yes   no   

         contact month day_of_week  ...  campaign  

In [ ]:
# Removing extra spaces
df = df.apply(lambda x: x.str.strip() if x.dtype == 'object' else x)

In [ ]:
# Changing columns name
df = df.rename(columns={
    'y'             : 'subscribed',
    'duration'      : 'call_duration_sec',
    'campaign'      : 'contacts_this_campaign',
    'pdays'         : 'days_since_last_contact',
    'previous'      : 'contacts_prev_campaign',
    'poutcome'      : 'prev_campaign_outcome',
    'emp.var.rate'  : 'emp_var_rate',
    'cons.price.idx': 'cons_price_idx',
    'cons.conf.idx' : 'cons_conf_idx',
    'nr.employed'   : 'nr_employed',
    'day_of_week'   : 'contact_day'
})

print(df.columns.tolist())

['age', 'job', 'marital', 'education', 'default', 'housing', 'loan', 'contact', 'month', 'contact_day', 'call_duration_sec', 'contacts_this_campaign', 'days_since_last_contact', 'contacts_prev_campaign', 'prev_campaign_outcome', 'emp_var_rate', 'cons_price_idx', 'cons_conf_idx', 'euribor3m', 'nr_employed', 'subscribed']


In [ ]:
df['subscribed_num'] = df['subscribed'].map({'yes': 1, 'no': 0})

# Verify
print(df['subscribed_num'].value_counts())
print(f"\nOverall conversion rate: {round(df['subscribed_num'].mean() * 100, 2)}%")

subscribed_num
0    36537
1     4639
Name: count, dtype: int64

Overall conversion rate: 11.27%


In [ ]:
df['call_duration_min'] = round(df['call_duration_sec'] / 60, 2)

# Verify
print(df[['call_duration_sec', 'call_duration_min']].head())

   call_duration_sec  call_duration_min
0                261               4.35
1                149               2.48
2                226               3.77
3                151               2.52
4                307               5.12


In [ ]:
bins = [0, 30, 45, 60, 100]
labels = ['18-30', '31-45', '46-60', '60+']

df['age_group'] = pd.cut(df['age'], bins=bins, labels=labels, right=True)

# Verify
print(df['age_group'].value_counts().sort_index())

age_group
18-30     7381
31-45    21967
46-60    10919
60+        909
Name: count, dtype: int64


In [ ]:
month_map = {
    'jan':1, 'feb':2, 'mar':3, 'apr':4,
    'may':5, 'jun':6, 'jul':7, 'aug':8,
    'sep':9, 'oct':10, 'nov':11, 'dec':12
}

df['month_num'] = df['month'].map(month_map)

# Verify
print(df[['month', 'month_num']].drop_duplicates().sort_values('month_num'))

      month  month_num
27690   mar          3
27972   apr          4
0       may          5
7763    jun          6
12137   jul          7
18822   aug          8
37887   sep          9
23997   oct         10
24064   nov         11
27680   dec         12


In [ ]:
month_map = {
    'jan':1, 'feb':2, 'mar':3, 'apr':4,
    'may':5, 'jun':6, 'jul':7, 'aug':8,
    'sep':9, 'oct':10, 'nov':11, 'dec':12
}

df['month_num'] = df['month'].map(month_map)

# Verify
print(df[['month', 'month_num']].drop_duplicates().sort_values('month_num'))

      month  month_num
27690   mar          3
27972   apr          4
0       may          5
7763    jun          6
12137   jul          7
18822   aug          8
37887   sep          9
23997   oct         10
24064   nov         11
27680   dec         12


In [ ]:
print("=== FINAL DATASET SUMMARY ===")
print(f"Shape: {df.shape}")
print(f"\nMissing values:\n{df.isnull().sum()}")
print(f"\nConversion rate: {round(df['subscribed_num'].mean()*100, 2)}%")
print(f"\nAge group distribution:\n{df['age_group'].value_counts().sort_index()}")
print(f"\nData types:\n{df.dtypes}")

=== FINAL DATASET SUMMARY ===
Shape: (41176, 25)

Missing values:
age                        0
job                        0
marital                    0
education                  0
default                    0
housing                    0
loan                       0
contact                    0
month                      0
contact_day                0
call_duration_sec          0
contacts_this_campaign     0
days_since_last_contact    0
contacts_prev_campaign     0
prev_campaign_outcome      0
emp_var_rate               0
cons_price_idx             0
cons_conf_idx              0
euribor3m                  0
nr_employed                0
subscribed                 0
subscribed_num             0
call_duration_min          0
age_group                  0
month_num                  0
dtype: int64

Conversion rate: 11.27%

Age group distribution:
age_group
18-30     7381
31-45    21967
46-60    10919
60+        909
Name: count, dtype: int64

Data types:
age                           int64
j

## Process Log — Cleaning Decisions

1.  Loaded raw file and created working copy — raw file never modified
2.  Ran df.info(), df.describe(), df.shape — initial profiling complete
3.  Checked missing values — zero nulls found across all 21 columns
4.  Found 12 duplicate rows — dropped and saved back to df
5.  Stripped leading/trailing whitespace from all string columns
6.  Checked unknown values — found in poutcome, contact, job, education, marital
7.  Replaced 'unknown' in poutcome → 'no_prior_contact'
8.  Replaced 'unknown' in contact → 'unrecorded'
9.  Kept 'unknown' as category in job, education, marital — too many to drop safely
10. Renamed all columns for Power BI readability
11. Encoded subscribed → subscribed_num (yes=1, no=0)
12. Converted call_duration_sec → call_duration_min
13. Binned age into 4 groups: 18-30, 31-45, 46-60, 60+
14. Mapped month names → month_num for correct Power BI sorting
15. Final validation — no nulls, clean dtypes, shape confirmed


# Analysis Phase

In [ ]:
# @title
# Overall conversion rate
total = len(df)
converted = df['subscribed_num'].sum()
not_converted = total - converted
rate = round(df['subscribed_num'].mean() * 100, 2)

print(f"Total customers contacted: {total}")
print(f"Converted: {converted}")
print(f"Not converted: {not_converted}")
print(f"Conversion rate: {rate}%")

Total customers contacted: 41176
Converted: 4639
Not converted: 36537
Conversion rate: 11.27%


In [ ]:
# @title
# Conversion rate by job
job_conv = df.groupby('job')['subscribed_num'].agg(['mean', 'count'])
job_conv['mean'] = round(job_conv['mean'] * 100, 2)
job_conv.columns = ['conversion_rate_%', 'total_contacts']
job_conv = job_conv.sort_values('conversion_rate_%', ascending=False)
print("=== CONVERSION BY JOB ===")
print(job_conv)

# Conversion rate by education
edu_conv = df.groupby('education')['subscribed_num'].agg(['mean', 'count'])
edu_conv['mean'] = round(edu_conv['mean'] * 100, 2)
edu_conv.columns = ['conversion_rate_%', 'total_contacts']
edu_conv = edu_conv.sort_values('conversion_rate_%', ascending=False)
print("\n=== CONVERSION BY EDUCATION ===")
print(edu_conv)

# Conversion rate by age group
age_conv = df.groupby('age_group')['subscribed_num'].agg(['mean', 'count'])
age_conv['mean'] = round(age_conv['mean'] * 100, 2)
age_conv.columns = ['conversion_rate_%', 'total_contacts']
print("\n=== CONVERSION BY AGE GROUP ===")
print(age_conv)

# Conversion rate by marital status
mar_conv = df.groupby('marital')['subscribed_num'].agg(['mean', 'count'])
mar_conv['mean'] = round(mar_conv['mean'] * 100, 2)
mar_conv.columns = ['conversion_rate_%', 'total_contacts']
mar_conv = mar_conv.sort_values('conversion_rate_%', ascending=False)
print("\n=== CONVERSION BY MARITAL STATUS ===")
print(mar_conv)

=== CONVERSION BY JOB ===
               conversion_rate_%  total_contacts
job                                             
student                    31.43             875
retired                    25.26            1718
unemployed                 14.20            1014
admin.                     12.97           10419
management                 11.22            2924
unknown                    11.21             330
technician                 10.83            6739
self-employed              10.49            1421
housemaid                  10.00            1060
entrepreneur                8.52            1456
services                    8.14            3967
blue-collar                 6.90            9253

=== CONVERSION BY EDUCATION ===
                     conversion_rate_%  total_contacts
education                                             
illiterate                       22.22              18
unknown                          14.51            1730
university.degree                13

/tmp/ipykernel_5103/3163540346.py:19: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.
  age_conv = df.groupby('age_group')['subscribed_num'].agg(['mean', 'count'])


In [ ]:
# @title
# Conversion rate by number of contacts
contact_conv = df.groupby('contacts_this_campaign')['subscribed_num'].agg(['mean', 'count'])
contact_conv['mean'] = round(contact_conv['mean'] * 100, 2)
contact_conv.columns = ['conversion_rate_%', 'total_contacts']
print("=== CONVERSION BY NUMBER OF CONTACTS ===")
print(contact_conv.head(10))

# Average call duration — converted vs not
duration_analysis = df.groupby('subscribed')['call_duration_min'].agg(['mean', 'median'])
duration_analysis.columns = ['avg_duration_min', 'median_duration_min']
print("\n=== CALL DURATION VS CONVERSION ===")
print(round(duration_analysis, 2))

# Conversion by contact method
contact_method = df.groupby('contact')['subscribed_num'].agg(['mean', 'count'])
contact_method['mean'] = round(contact_method['mean'] * 100, 2)
contact_method.columns = ['conversion_rate_%', 'total_contacts']
print("\n=== CONVERSION BY CONTACT METHOD ===")
print(contact_method)

=== CONVERSION BY NUMBER OF CONTACTS ===
                        conversion_rate_%  total_contacts
contacts_this_campaign                                   
1                                   13.04           17634
2                                   11.46           10568
3                                   10.75            5340
4                                    9.40            2650
5                                    7.50            1599
6                                    7.66             979
7                                    6.04             629
8                                    4.25             400
9                                    6.01             283
10                                   5.33             225

=== CALL DURATION VS CONVERSION ===
            avg_duration_min  median_duration_min
subscribed                                       
no                      3.68                 2.73
yes                     9.22                 7.48

=== CONVERSION BY CONTACT

In [ ]:
# @title
# Conversion by month
month_conv = df.groupby(['month_num', 'month'])['subscribed_num'].agg(['mean', 'count'])
month_conv['mean'] = round(month_conv['mean'] * 100, 2)
month_conv.columns = ['conversion_rate_%', 'total_contacts']
month_conv = month_conv.sort_values('month_num')
print("=== CONVERSION BY MONTH ===")
print(month_conv)

# Conversion by day of week
day_conv = df.groupby('contact_day')['subscribed_num'].agg(['mean', 'count'])
day_conv['mean'] = round(day_conv['mean'] * 100, 2)
day_conv.columns = ['conversion_rate_%', 'total_contacts']
print("\n=== CONVERSION BY DAY OF WEEK ===")
print(day_conv)

=== CONVERSION BY MONTH ===
                 conversion_rate_%  total_contacts
month_num month                                   
3         mar                50.55             546
4         apr                20.49            2631
5         may                 6.44           13767
6         jun                10.51            5318
7         jul                 9.04            7169
8         aug                10.61            6176
9         sep                44.91             570
10        oct                43.93             717
11        nov                10.15            4100
12        dec                48.90             182

=== CONVERSION BY DAY OF WEEK ===
             conversion_rate_%  total_contacts
contact_day                                   
fri                      10.81            7826
mon                       9.95            8512
thu                      12.11            8618
tue                      11.79            8086
wed                      11.67            8

In [ ]:
# Housing loan vs conversion
housing_conv = df.groupby('housing')['subscribed_num'].agg(['mean', 'count'])
housing_conv['mean'] = round(housing_conv['mean'] * 100, 2)
housing_conv.columns = ['conversion_rate_%', 'total_contacts']
print("\n=== HOUSING LOAN VS CONVERSION ===")
print(housing_conv)

# Personal loan vs conversion
loan_conv = df.groupby('loan')['subscribed_num'].agg(['mean', 'count'])
loan_conv['mean'] = round(loan_conv['mean'] * 100, 2)
loan_conv.columns = ['conversion_rate_%', 'total_contacts']
print("\n=== PERSONAL LOAN VS CONVERSION ===")
print(loan_conv)


=== HOUSING LOAN VS CONVERSION ===
         conversion_rate_%  total_contacts
housing                                   
no                   10.88           18615
unknown              10.81             990
yes                  11.62           21571

=== PERSONAL LOAN VS CONVERSION ===
         conversion_rate_%  total_contacts
loan                                      
no                   11.34           33938
unknown              10.81             990
yes                  10.93            6248


In [ ]:
# Saving the clean dataset
df.to_csv('bank_telemarketing_cleaned.csv', index=False)
print("Cleaned data saved to 'bank_telemarketing_cleaned.csv'")

Cleaned data saved to 'bank_telemarketing_cleaned.csv'


## Key Insights — Pre-Visualization

1. Overall conversion rate is 11.27% — below industry average for
   telemarketing (typically 15-20%), suggesting targeting needs improvement.

2.  Retired and student segments are highest performing.

3. Conversion drops sharply after 5+ contact attempts.
   Diminishing returns are significant — contacts beyond 5 waste resources.

4. Cellular contact outperforms telephone by around 10 percentage.

5. March , December and September show highest conversion rates.
   May has highest volume but lowest conversion — misaligned effort.


# Share


### Dashboard Summary
Tool: Power BI Desktop
File: bank_campaign_dashboard_v1.pbix
Pages: 4 (Overview, Customer Profile,
       Campaign Analysis, Financial Segments)

### Key Findings

Finding 1 — Demographic Drivers
The 60+ age group converts at 40%+ — nearly 4x the 11.27%
campaign average. Students follow at 31%. Blue-collar,
services, and entrepreneur segments fall below 10%.

Finding 2 — Campaign Behavior
Conversion peaks at the first contact attempt and drops
sharply after 5 calls. Cellular contact converts at 14%
vs 5% for landline — nearly 3x more effective.

Finding 3 — Previous Campaign Impact
Customers with prior successful engagement convert at
20–60%. This is the single strongest predictor of
subscription in the entire dataset.

# Act

### Executive Summary

SITUATION
41,188 customers contacted. Overall conversion rate: 11.27%

RECOMMENDATION 1 — PRIORITIZE BOOKEND DEMOGRAPHICS
Target 60+ and student segments. Suppress outreach to
blue-collar and services unless prior engagement exists.

RECOMMENDATION 2 — IMPLEMENT 5-CALL HARD CAP
Cap outbound calls at 5 attempts per customer. Every call
beyond that inflates CPA with near-zero conversion return.

RECOMMENDATION 3 — MOBILE-FIRST STRATEGY
Prioritize cellular numbers in auto-dialer before landlines.
This alone could lift conversion without adding new contacts.

RECOMMENDATION 4 — WARM LEAD ROUTING
Route previously engaged customers to senior agents first.
These leads convert at up to 60% — highest value segment.

PROJECTED IMPACT
Implementing all four recommendations on a future 10,000
customer campaign could lift conversion from 11.27% to
18–22% while reducing cost-per-acquisition by 30–40%.